docs/dev/stream/state/custom_serialization.md

```diff
@@ -66,125 +69,166 @@ checkpointedState = getRuntimeContext.getListState(descriptor)
+  `TypeSerializer#resolveSchemaCompatibility` to check for schema compatibility.
+ 4. **Migrate state bytes in backend from schema _A_ to schema _B_**
+  - If the compatibility resolution reflects that the schema has changed and migration is possible, schema migration is 
+  performed. The prior state serializer which recognizes schema _A_ will be obtained from the serializer snapshot, via
+   `TypeSerializerSnapshot#restoreSerializer()`, and is used to deserialize state bytes to objects, which in turn
+   is re-written again with the new serializer, which recognizes schema _B_ to complete the migration.
+  - If the resolution signals incompatibility, then the state access fails with an exception.
+ 
+#### Heap state backends (e.g. `MemoryStateBackend`, `FsStateBackend`)
+
+ 1. **Register new state with a state serializer that has schema _A_**
+  - the registered `TypeSerializer` is maintained by the state backend.
+ 2. **Take a savepoint, serializing all state with schema _A_**
+  - The serializer snapshot is extracted via the `TypeSerializer#snapshotConfiguration` method.
+  - The serializer snapshot is written to the savepoint.
+  - State objects are now serialized to the savepoint, written in schema _A_.
+ 3. **On restore, deserialize state into objects in heap**
+  - The previous state serializer's snapshot is restored.
+  - The prior serializer, which recognizes schema _A_, is obtained from the serializer snapshot, via
+  `TypeSerializerSnapshot#restoreSerializer()`, and is used to deserialize state bytes to objects.
+  - From now on, all of the state is already deserialized.
+ 4. **Restored execution re-accesses prior state with new state serializer that has schema _B_**
+  - Upon receiving the new serializer, it is provided to the restored prior serializer's snapshot via the
+  `TypeSerializer#resolveSchemaCompatibility` to check for schema compatibility.
+  - If the compatibility check signals that migration is required, nothing happens in this case since for
+   heap backends, all state is already deserialized into objects.
+  - If the resolution signals incompatibility, then the state access fails with an exception.
+ 5. **Take another savepoint, serializing all state with schema _B_**
+  - Same as step 2., but now state bytes are all in schema _B_.
+
+## Implementation notes and best practices
+
+#### 1. Flink restores serializer snapshots by instantiating them with their classname
+
+A serializer's snapshot, being the single source of truth for how a registered state was serialized, serves as an
+entrypoint to reading state in savepoints. In order to be able to restore and access previous state, the prior state
+serializer's snapshot must be able to be restored.
+
+Flink restores serializer snapshots by first instantiating the `TypeSerializerSnapshot` with its classname (written
+along with the snapshot bytes). Therefore, to avoid being subject to unintended classname changes,
```

Comment:
I would list those requirements in a bulleted list. I find lists easier to follow, plus one can treat them as check lists.